src/test/java/io/kamax/matrix/client/regular/MatrixHttpClientTest.java

```diff
@@ -39,6 +50,95 @@ public MatrixHttpClientTest() throws URISyntaxException {
     }
 
     @Test
+    public void loginWithDeviceId() throws URISyntaxException {
+        stubFor(post(urlEqualTo(loginUrl))
+                .withRequestBody(equalToJson("{\"type\": \"m.login.password\"," + //
+                        "\"user\": \"" + user.getLocalPart() + "\"," + //
+                        "\"password\": \"" + password + "\"," + //
+                        "\"device_id\": \"" + deviceId + "\"}"))
+                .willReturn(aResponse().withStatus(200)
+                        .withBody("{\"user_id\": \"" + user.getId() + "\"," + //
+                                "\"access_token\": \"" + testToken + "\"," + //
+                                "\"home_server\": \"" + new MatrixHomeserver(domain, baseUrl) + "\"," + //
```

Comment:
we rely on `toString()` here, which might be a bad idea. [The spec](https://matrix.org/speculator/spec/HEAD/client_server/unstable.html#post-matrix-client-r0-login) says it should be the hostname, but the test is not clear on what it will be - Full URL or domain? or something else? Best to be clear.